DAO/src/lib.rs

```diff
@@ -0,0 +1,698 @@
+        let member = self.members.get_mut(&member_id).unwrap();
+
+        match vote {
+            Vote::Yes => {
+                proposal.yes_votes = proposal.yes_votes.saturating_add(member.shares);
+                if self.total_shares > proposal.max_total_shares_at_yes_vote {
+                    proposal.max_total_shares_at_yes_vote = self.total_shares;
+                }
+                // it is necessary to save the highest id of the proposal - must be processed for member to ragequit
+                if member.highest_index_yes_vote < proposal_id {
+                    member.highest_index_yes_vote = proposal_id;
+                }
+            }
+            Vote::No => {
+                proposal.no_votes = proposal.no_votes.saturating_add(member.shares);
+            }
+        }
+        proposal.votes_by_member.insert(msg::source(), vote.clone());
+
+        msg::reply(
+            DaoEvent::SubmitVote {
+                account: msg::source(),
+                proposal_id,
+                vote,
+            },
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    // The proposal processing after the proposal completes during the grace period.
+    // If the proposal is accepted, the tribute tokens are deposited into the contract and new shares are minted and issued to the applicant.
+    // If the proposal is rejected, the tribute tokens are returned to the applicant.
+    // Requirements:
+    // * The previous proposal must be processed
+    // * The proposal must exist, be ready for processing
+    // * The proposal must not be cancelled, aborted or already be processed
+    // Arguments:
+    // * `proposal_id`: the proposal ID
+    async fn process_proposal(&mut self, proposal_id: u128) {
+        match self.proposals.get(&proposal_id) {
+            Some(proposal) => {
+                if proposal_id > 0 && !self.proposals.get(&(&proposal_id - 1)).unwrap().processed {
+                    panic!("Previous proposal must be processed");
+                }
+                if proposal.processed || proposal.cancelled || proposal.aborted {
+                    panic!("Proposal has already been processed, cancelled or aborted");
+                }
+                if exec::block_timestamp()
+                    < proposal.starting_period
+                        + self.voting_period_length
+                        + self.grace_period_length
+                {
+                    panic!("Proposal is not ready to be processed");
+                }
+            }
+            None => {
+                panic!("proposal does not exist");
+            }
+        }
+
+        let mut proposal = self.proposals.get(&proposal_id).unwrap().clone();
+        proposal.processed = true;
+        proposal.did_pass = proposal.yes_votes > proposal.no_votes
+            && proposal.yes_votes * 10000 / self.total_shares >= proposal.quorum
```

Comment:
suggestion : `proposal.yes_votes * 10000 / self.total_shares` this seems to be used at few other places too. Better to create function. Also 10000 can be made named constant.

Reply:
That calculation is used only in that place

Reply:
Isn't line number 350 and 452 trying to check if quorum is reached or not. But its upto you to create a function or not.
On line 350 its * 10000 and on 452 its * 1000 . If those numbers should be same then better create a cons value.